geomesa-feature/src/main/scala/org/locationtech/geomesa/feature/kryo/KryoSimpleFeature.scala

```diff
@@ -0,0 +1,132 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.locationtech.geomesa.feature.kryo
+
+import java.util.{Collection => JCollection, List => JList}
+
+import com.vividsolutions.jts.geom.Geometry
+import org.geotools.feature.`type`.{AttributeDescriptorImpl, Types}
+import org.geotools.feature.{AttributeImpl, GeometryAttributeImpl}
+import org.geotools.filter.identity.FeatureIdImpl
+import org.geotools.geometry.jts.ReferencedEnvelope
+import org.geotools.util.Converters
+import org.opengis.feature.`type`.{AttributeDescriptor, Name}
+import org.opengis.feature.simple.{SimpleFeature, SimpleFeatureType}
+import org.opengis.feature.{GeometryAttribute, Property}
+import org.opengis.geometry.BoundingBox
+
+import scala.collection.JavaConversions._
+
+/**
+ * Simple feature implementation optimized to instantiate from kryo serialization
+ *
+ * @param initialId
+ * @param sft
+ * @param initialValues if provided, must already be converted into the appropriate types
+ */
+class KryoSimpleFeature(initialId: String, sft: SimpleFeatureType, initialValues: Array[AnyRef] = null)
+    extends SimpleFeature {
+
+  val featureId = new FeatureIdImpl(initialId)
+  val values = if (initialValues == null) Array.ofDim[AnyRef](sft.getAttributeCount) else initialValues
+
+  lazy private[this] val userData  = collection.mutable.HashMap.empty[AnyRef, AnyRef]
+  lazy private[this] val geomDesc  = sft.getGeometryDescriptor
+  lazy private[this] val geomIndex = if (geomDesc == null) -1 else sft.indexOf(geomDesc.getLocalName)
+
+  override def getFeatureType = sft
+  override def getType = sft
+  override def getIdentifier = featureId
+  override def getID = featureId.getID // this needs to reference the featureId, as it can be updated
+  override def getName: Name = sft.getName
+  override def getUserData = userData
+
+  override def getAttribute(name: Name) = getAttribute(name.getLocalPart)
+  override def getAttribute(name: String) = {
+    val index = sft.indexOf(name)
+    if (index == -1) null else getAttribute(index)
+  }
+  override def getAttribute(index: Int) = values(index)
+
+  override def setAttribute(name: Name, value: Object) = setAttribute(name.getLocalPart, value)
+  override def setAttribute(name: String, value: Object) = {
+    val index = sft.indexOf(name)
+    if (index == -1) {
+      throw new IllegalArgumentException(s"Attribute $name does not exist in type $sft")
+    }
+    setAttribute(index, value)
+  }
+  override def setAttribute(index: Int, value: Object) = {
+    val binding = sft.getDescriptor(index).getType.getBinding
+    values(index) = Converters.convert(value, binding).asInstanceOf[AnyRef]
+  }
+
+  // following methods delegate to setAttribute to get type conversion
+  override def setAttributes(vals: JList[Object]) =
+    vals.zipWithIndex.foreach { case (v, i) => setAttribute(i, v) }
```

Comment:
From personal experience, I'm against zipWithIndex.  It instantiates additional objects for this control structure which slows things down. 

Since we are pushing for speed, I'd suggest trying a more 'Java' approach to the iteration.